### PR TITLE
Feat/delegate call

### PR DESCRIPTION
**Pull request by Eikix**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves: #394 #393

## What is the new behavior?

- add delegate call 
- add staticcall
- ⚠️ rebased on staticcall PR (has its commits as well)
- make addresses in machine and execution context an Address struct rather than an evm address

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: crates/contracts/src/kakarot_core/kakarot.cairo [hunk 1]
```diff
@@ -316,6 +316,27 @@ mod KakarotCore {
                         Option::Some(account) => account.bytecode()?,
                         Option::None => Default::default().span(),
                     };
+
+                    // `from` address should be an EOA, i.e. a deployed Starknet account
+                    // checking invariants here:
```
Comment (Eikix) [inline crates/contracts/src/kakarot_core/kakarot.cairo:321]: don't have a strong opinion about invariants here

File: crates/contracts/src/kakarot_core/kakarot.cairo [hunk 2]
```diff
@@ -317,6 +317,27 @@ mod KakarotCore {
...
+                    // - `from` MUST not be an uninitialized account
+                    // - `from` MUST not be a contract account
+                    let from_starknet_address = match self.address_registry(from) {
```
Comment (enitrat) [inline crates/contracts/src/kakarot_core/kakarot.cairo:325]: We should have these checks outside of `handle_call`

 - In `eth_call`, verify that `from` is a an address that is registered and has a corresponding starknet address
 - In `eth_send_transaction`, verify that `from` is the transaction origin, and that it's registered and has a corresponding SN address

Comment (enitrat) [reply crates/contracts/src/kakarot_core/kakarot.cairo:325]: Handle call could therefore receive one extra parameter which would be `starknet_address`

File: crates/evm/src/execution.cairo [hunk 3]
```diff
@@ -31,8 +31,8 @@ use utils::helpers::compute_starknet_address;
 /// *   The created contracts
 /// *   The events emitted
 fn execute(
-    origin: EthAddress,
-    target: EthAddress,
+    origin: Address,
+    address: Address,
```
Comment (enitrat) [inline crates/evm/src/execution.cairo:35]: semantically I think `target` has more meaning, especially given that we have a lot of variables named `address` and I'm often a bit lost when trying to remember which one corresponds to what

File: crates/evm/src/execution.cairo [hunk 4]
```diff
@@ -63,38 +63,7 @@ fn execute(
     // Initiate the Machine with the root context
     let mut machine: Machine = MachineCurrentContextTrait::new(ctx);
 
-    // Handle value transfers
-    let result_sender = machine.state.get_account(origin);
```
Comment (enitrat) [inline crates/evm/src/execution.cairo:67]: Do you see any cases where this might cause a problem to not fetch the account from state first?

Comment (Eikix) [reply crates/evm/src/execution.cairo:67]: afai see, the get_account is solely used to get the sender's Address struct, here we provide it before hand

Comment (enitrat) [reply crates/evm/src/execution.cairo:67]: ok

**Review by enitrat — APPROVED** [review]
image

**Pull request merged**